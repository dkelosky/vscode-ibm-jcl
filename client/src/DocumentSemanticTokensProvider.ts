import * as vscode from "vscode";

const tokenTypes = new Map<string, number>();
const tokenModifiers = new Map<string, number>();

// This maps tokensLegend defaults to TextMate Grammers
// https://github.com/microsoft/vscode/blob/be0aca7188ec6a76e7c2379758c0fbc1e9c21f7b/src/vs/platform/theme/common/tokenClassificationRegistry.ts#L372-L408

export const legend = (() => {
    const tokenTypesLegend = [
        "comment", "string", "keyword", "number", "regexp", "operator", "namespace",
        "type", "struct", "class", "interface", "enum", "typeParameter", "function",
        "member", "macro", "variable", "parameter", "property", "label"
    ];
    tokenTypesLegend.forEach((tokenType, index) => tokenTypes.set(tokenType, index));

    // these need editor config to show
    const tokenModifiersLegend = [
        "declaration", "documentation", "readonly", "static", "abstract", "deprecated",
        "modification", "async"
    ];
    tokenModifiersLegend.forEach((tokenModifier, index) => tokenModifiers.set(tokenModifier, index));

    return new vscode.SemanticTokensLegend(tokenTypesLegend, tokenModifiersLegend);
})();

export function activate(context: vscode.ExtensionContext): void {
    context.subscriptions.push(vscode.languages
        .registerDocumentSemanticTokensProvider({ language: "jcl" }, new DocumentSemanticTokensProvider(), legend));
}

interface IParsedToken {
    line: number;
    startCharacter: number;
    length: number;
    tokenType: string;
    tokenModifiers: string[];
}

export class DocumentSemanticTokensProvider implements vscode.DocumentSemanticTokensProvider {
    async provideDocumentSemanticTokens(document: vscode.TextDocument, token: vscode.CancellationToken):
        Promise<vscode.SemanticTokens> {
        const allTokens = this._parseText(document.getText());
        const builder = new vscode.SemanticTokensBuilder();
        allTokens.forEach((token) => {
            builder.push(token.line, token.startCharacter, token.length,
                this._encodeTokenType(token.tokenType), this._encodeTokenModifiers(token.tokenModifiers));
        });
        return builder.build();
    }

    private _encodeTokenType(tokenType: string): number {
        if (tokenTypes.has(tokenType)) {
            return tokenTypes.get(tokenType)!;
        } else if (tokenType === "notInLegend") {
            return tokenTypes.size + 2;
        }
        return 0;
    }

    private _encodeTokenModifiers(strTokenModifiers: string[]): number {
        let result = 0;
        for (let i = 0; i < strTokenModifiers.length; i++) {
            const tokenModifier = strTokenModifiers[i];
            if (tokenModifiers.has(tokenModifier)) {
                result = result | (1 << tokenModifiers.get(tokenModifier)!);
            } else if (tokenModifier === "notInLegend") {
                result = result | (1 << tokenModifiers.size + 2);
            }
        }
        return result;
    }

    private _parseText(text: string): IParsedToken[] {
        let r: IParsedToken[] = [];
        let lines = text.split(/\r\n|\r|\n/);
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i];
            let currentOffset = 0;
            let openOffset = -1;
            let closeOffset = -1;

            let tokens = line.replace(/\s+/g, " ").split(" ");
            tokens = tokens.filter((entry) => entry !== "");

            // where we have // s
            if (tokens.length > 1) {

                if (tokens[1].indexOf("IF") > -1) {
                    openOffset = line.indexOf(" IF");


                    if ()
                    }

            }

            for (let j = 0; j < tokens.length; j++) {

                console.log(`${tokens[j]}`)
            }

            if (openOffset === -1) {
                break;
            }
            // const openOffset = line.indexOf("[", currentOffset);
            // if (openOffset === -1) {
            //     break;
            // }
            // const closeOffset = line.indexOf("]", openOffset);
            // if (closeOffset === -1) {
            //     break;
            // }
            // let tokenData = this._parseTextToken(line.substring(openOffset + 1, closeOffset));
            r.push({
                line: i,
                startCharacter: openOffset + 1,
                length: closeOffset - openOffset - 1,
                tokenType: tokenData.tokenType,
                tokenModifiers: tokenData.tokenModifiers
            });
            // currentOffset = closeOffset;
            // } while (true);
        }
        return r;
    }


    private _parseTextToken(text: string): { tokenType: string; tokenModifiers: string[]; } {
        let parts = text.split(".");
        return {
            tokenType: parts[0],
            tokenModifiers: parts.slice(1)
        };
    }
}
