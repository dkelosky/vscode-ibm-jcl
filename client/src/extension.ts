/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */

import * as path from "path";
import { workspace, ExtensionContext, window, commands } from "vscode";
import TelemetryReporter  from "vscode-extension-telemetry";

import {
    LanguageClient,
    LanguageClientOptions,
    ServerOptions,
    TransportKind
} from "vscode-languageclient";

let client: LanguageClient;

// all events will be prefixed with this event name
const extensionId = 'ibm-jcl';

// extension version will be reported as a property with each event 
const extensionVersion = '0.11.0';

// the application insights key (also known as instrumentation key)
const key = '74997a52-f83f-42b1-bd75-586c155a8fc5';

// telemetry reporter 
let reporter: TelemetryReporter;

export function activate(context: ExtensionContext) {

    reporter = new TelemetryReporter(extensionId, extensionVersion, key);
    // ensure it gets property disposed
    context.subscriptions.push(reporter);

    const disposable = commands.registerCommand('extension.helloWorld', () => {
		// The code you place here will be executed every time your command is executed

		// Display a message box to the user
        window.showInformationMessage('Hello World!');
        
        reporter.sendTelemetryEvent('sampleEvent', { 'stringProp': 'some string' }, { 'numericMeasure': 123 });
	});

    // The server is implemented in node
    const serverModule = context.asAbsolutePath(
        path.join("server", "out", "server.js")
    );
    // The debug options for the server
    // --inspect=6009: runs the server in Node"s Inspector mode so VS Code can attach to the server for debugging
    const debugOptions = { execArgv: ["--nolazy", "--inspect=6010"] };

    // If the extension is launched in debug mode then the debug server options are used
    // Otherwise the run options are used
    const serverOptions: ServerOptions = {
        run: { module: serverModule, transport: TransportKind.ipc },
        debug: {
            module: serverModule,
            transport: TransportKind.ipc,
            options: debugOptions
        },
    };

    // Options to control the language client
    const clientOptions: LanguageClientOptions = {
        // Register the server for plain text documents
        documentSelector: [{ scheme: "file", language: "jcl" }],
        synchronize: {
            // Notify the server about file changes to ".clientrc files contained in the workspace
            fileEvents: workspace.createFileSystemWatcher("**/.jcl")
        }
    };

    // Create the language client and start the client.
    client = new LanguageClient(
        "jclServer",
        "JCL Server",
        serverOptions,
        clientOptions
    );

    // Start the client. This will also launch the server
    client.start();

    // TODO(Kelosky): refine errors
    // context.subscriptions.push(
    //     languages.registerDocumentSemanticTokensProvider({ language: "jcl"}, new DocumentSemanticTokensProvider(), legend));
}

export function deactivate(): Thenable<void> | undefined {
    // This will ensure all pending events get flushed
    reporter.dispose();

    if (!client) {
        return undefined;
    }
    return client.stop();
}