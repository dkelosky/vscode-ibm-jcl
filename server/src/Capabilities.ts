import { InitializeParams } from "vscode-languageserver";

export class Capabilities {
    
    public hasConfigurationCapability: boolean;
    public hasWorkspaceFolderCapability: boolean;
    public hasDiagnosticRelatedInformationCapability: boolean;

    constructor() {
        this.hasConfigurationCapability = false;
        this.hasWorkspaceFolderCapability = false;
        this.hasDiagnosticRelatedInformationCapability = false;
    }

    public init(params: InitializeParams): void {
        const capabilities = params.capabilities;

        // Does the client support the `workspace/configuration` request?
        // If not, we will fall back using global settings
        this.hasConfigurationCapability = !!(
            capabilities.workspace && !!capabilities.workspace.configuration
        );
        this.hasWorkspaceFolderCapability = !!(
            capabilities.workspace && !!capabilities.workspace.workspaceFolders
        );
        this.hasDiagnosticRelatedInformationCapability = !!(
            capabilities.textDocument &&
            capabilities.textDocument.publishDiagnostics &&
            capabilities.textDocument.publishDiagnostics.relatedInformation
        );
    }
}
