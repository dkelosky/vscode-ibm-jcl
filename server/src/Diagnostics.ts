import { TextDocument, Diagnostic, DiagnosticSeverity } from "vscode-languageserver";

export class Diagnostics {
    public static readonly MAX_LEN = 80;

    public static checkLengths(document: TextDocument, diagnostics: Diagnostic[]): void {

        const text = document.getText();

        const lines = text.split(/\r?\n/);
        for (let i = 0; i < lines.length - 1; i++) {

            if (lines[i].length > Diagnostics.MAX_LEN) {
                const diagnostic: Diagnostic = {
                    severity: DiagnosticSeverity.Error,
                    range: {
                        start: { line: i, character: Diagnostics.MAX_LEN },
                        end: { line: i, character: lines[i].length - 1 },
                    },
                    message: `Exceeded ${Diagnostics.MAX_LEN} characters`,
                    source: "jcl"
                };

                diagnostics.push(diagnostic);
            }

        }
    }
}
