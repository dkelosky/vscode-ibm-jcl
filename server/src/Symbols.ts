import { SymbolInformation, TextDocument, SymbolKind } from "vscode-languageserver";

export class Symbols {

    public static parse(document: TextDocument): SymbolInformation[] {
        const symbols: SymbolInformation[] = [];
        const lines = document.getText().split("\n");

        for (let i = 0; i < lines.length - 1; i++) {

            if (lines[i][0] === "/" && lines[i][1] === "/" && lines[i][2] !== "*" && lines[i][2] !== " ") {

                // compress multiple spaces to a single space
                let tokenizedLine = lines[i].replace(/\s+/g, " ").split(" ");

                // remove blank entries
                tokenizedLine = tokenizedLine.filter((entry) => entry !== "");

                if (tokenizedLine.length > 0) {

                    const entry: SymbolInformation = {
                        name: tokenizedLine[0],
                        kind: SymbolKind.Constant,
                        location: {
                            uri: document.uri,
                            range: {
                                start: { line: i, character: 0 },
                                end: { line: i, character: tokenizedLine[0].length - 1 }
                            }
                        }
                    };

                    symbols.push(entry);
                }

            }
        }
        return symbols;
    }
}