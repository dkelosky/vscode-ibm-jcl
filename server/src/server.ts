/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */

import {
    createConnection,
    TextDocuments,
    ProposedFeatures,
    InitializeParams,
    DidChangeConfigurationNotification,
    SymbolKind,
    SymbolInformation,
    TextDocument,
    Diagnostic,
    DiagnosticSeverity
} from "vscode-languageserver";

const MAX_LEN = 80;

// Create a connection for the server. The connection uses Node's IPC as a transport.
// Also include all preview / proposed LSP features.
const connection = createConnection(ProposedFeatures.all);

// Create a simple text document manager. The text document manager
// supports full document sync only
const documents: TextDocuments = new TextDocuments();

let hasConfigurationCapability: boolean = false;
let hasWorkspaceFolderCapability: boolean = false;
let hasDiagnosticRelatedInformationCapability: boolean = false;

connection.onInitialize((params: InitializeParams) => {
    const capabilities = params.capabilities;

    // Does the client support the `workspace/configuration` request?
    // If not, we will fall back using global settings
    hasConfigurationCapability = !!(
        capabilities.workspace && !!capabilities.workspace.configuration
    );
    hasWorkspaceFolderCapability = !!(
        capabilities.workspace && !!capabilities.workspace.workspaceFolders
    );
    hasDiagnosticRelatedInformationCapability = !!(
        capabilities.textDocument &&
        capabilities.textDocument.publishDiagnostics &&
        capabilities.textDocument.publishDiagnostics.relatedInformation
    );

    return {
        capabilities: {
            textDocumentSync: documents.syncKind,
            documentSymbolProvider: true,
        }
    };
});

connection.onInitialized(() => {
    if (hasConfigurationCapability) {
        // Register for all configuration changes.
        connection.client.register(DidChangeConfigurationNotification.type, undefined);
    }
    if (hasWorkspaceFolderCapability) {
        connection.workspace.onDidChangeWorkspaceFolders((event) => {
            connection.console.log("Workspace folder change event received.");
        });
    }
});

// The example settings
interface IJclSettings {
    maxNumberOfProblems: number;
}

// The global settings, used when the `workspace/configuration` request is not supported by the client.
// Please note that this is not the case when using this server with the client provided in this example
// but could happen with other clients.
const defaultSettings: IJclSettings = { maxNumberOfProblems: 1000 };
let globalSettings: IJclSettings = defaultSettings;

// Cache the settings of all open documents
const documentSettings: Map<string, Thenable<IJclSettings>> = new Map();

connection.onDidChangeConfiguration((change) => {
    if (hasConfigurationCapability) {
        // Reset all cached document settings
        documentSettings.clear();
    } else {
        globalSettings =
            (change.settings.languageServerExample || defaultSettings);
    }

    // Revalidate all open text documents
    documents.all().forEach(validateTextDocument);
});

function getDocumentSettings(resource: string): Thenable<IJclSettings> {
    if (!hasConfigurationCapability) {
        return Promise.resolve(globalSettings);
    }
    let result = documentSettings.get(resource);
    if (!result) {
        result = connection.workspace.getConfiguration({
            scopeUri: resource,
            section: "jclServer"
        });
        documentSettings.set(resource, result);
    }
    return result;
}

// Only keep settings for open documents
documents.onDidClose(e => {
    documentSettings.delete(e.document.uri);
});

// The content of a text document has changed. This event is emitted
// when the text document first opened or when its content has changed.
documents.onDidChangeContent((change) => {
    validateTextDocument(change.document);
});

async function validateTextDocument(textDocument: TextDocument): Promise<void> {
    // In this simple example we get the settings for every validate run.
    const settings = await getDocumentSettings(textDocument.uri);

    // The validator creates diagnostics for all uppercase words length 2 and more
    const text = textDocument.getText();

    let problems = 0;
    const diagnostics: Diagnostic[] = [];

    // TODO(Kelosky): account for max number of problems
    //   while ((m = pattern.exec(text)) && problems < settings.maxNumberOfProblems) {
    // TODO(Kelosky): print diagnostics for blank lines

    const lines = text.split("\n");
    for (let i = 0; i < lines.length - 1; i++) {
        if (lines[i].length - 1 > MAX_LEN) {
            problems++;

            const diagnostic: Diagnostic = {
                severity: DiagnosticSeverity.Error,
                range: {
                    start: { line: i, character: MAX_LEN },
                    end: { line: i, character: lines[i].length - 1 },
                },
                message: `Exceeded ${MAX_LEN} characters`,
                source: "jcl"
            };
            // if (hasDiagnosticRelatedInformationCapability) {
            //     diagnostic.relatedInformation = [
            //         {
            //             location: {
            //                 uri: textDocument.uri,
            //                 range: Object.assign({}, diagnostic.range)
            //             },
            //             message: 'Spelling matters'
            //         },
            //         {
            //             location: {
            //                 uri: textDocument.uri,
            //                 range: Object.assign({}, diagnostic.range)
            //             },
            //             message: 'Particularly for names'
            //         }
            //     ];
            // }
            diagnostics.push(diagnostic);
        }

        // const tokenizedLine = lines[i].replace(/\s+/g, " ").split(" ");

        // if (tokenizedLine.length > 0) {

        //     let kind: SymbolKind = SymbolKind.Constant;

        //     // TODO(Kelosky): if LABEL MVC -> this fails to capture
        //     if (tokenizedLine[0]) {
        //         if (tokenizedLine[0] === "MVC") {
        //             kind = SymbolKind.Object;
        //         }

        //         if (!tokenizedLine[1]) {
        //             // error missing arguments
        //             const start = lines[i].indexOf(tokenizedLine[0]);
        //             const diagnostic: Diagnostic = {
        //                 severity: DiagnosticSeverity.Error,
        //                 range: {
        //                     start: { line: i, character: start },
        //                     end: { line: i, character: start + tokenizedLine[0].length },
        //                 },
        //                 message: `Argument(s) missing for ${tokenizedLine[0]}`,
        //                 source: "hlasm"
        //             };
        //         }
        //     }
        // }

    }

    // Send the computed diagnostics to VSCode.
    connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
}

connection.onDocumentSymbol((parm) => {

    const symbols: SymbolInformation[] = [];
    const document = documents.get(parm.textDocument.uri);
    if (!document) {
        return null;
    }

    const lines = document.getText().split("\n");
    for (let i = 0; i < lines.length - 1; i++) {

        // if space or * in column one, it's not a symbol
        if (lines[i][0] === "/" && lines[i][1] === "/" && lines[i][2] !== "*" && lines[i][2] !== " ") {

            // compress multiple spaces to a single space
            let  tokenizedLine = lines[i].replace(/\s+/g, " ").split(" ");

            // remove blank entries
            tokenizedLine = tokenizedLine.filter((entry) => entry !== "");

            if (tokenizedLine.length > 0) {

                let kind: SymbolKind = SymbolKind.Constant;

                if (tokenizedLine[1]) {
                    if (tokenizedLine[1] === "DSECT") {
                        kind = SymbolKind.Object;
                    }
                }

                const entry: SymbolInformation = {
                    name: tokenizedLine[0],
                    kind,
                    location: {
                        uri: parm.textDocument.uri,
                        range: {
                            start: { line: i, character: 0 },
                            end: { line: i, character: tokenizedLine[0].length - 1 }
                        }
                    }
                };

                symbols.push(entry);
            }

        }
    }
    return symbols;
});

// Make the text document manager listen on the connection
// for open, change and close text document events
documents.listen(connection);

// Listen on the connection
connection.listen();
