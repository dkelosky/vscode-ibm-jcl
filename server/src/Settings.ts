import { Connection } from "vscode-languageserver";
import { Capabilities } from "./Capabilities";

export interface IJclSettings {
    maxNumberOfProblems: number;
}

export class Settings {

    // The global settings, used when the `workspace/configuration` request is not supported by the client.
    // Please note that this is not the case when using this server with the client provided in this example
    // but could happen with other clients.
    public static readonly defaultSettings: IJclSettings = { maxNumberOfProblems: 1000 };
    public globalSettings: IJclSettings;

    // Cache the settings of all open documents
    public documentSettings: Map<string, Thenable<IJclSettings>>;

    constructor(private connection: Connection, private capabilities: Capabilities) {
        this.documentSettings = new Map();
        this.globalSettings = Settings.defaultSettings;
    }

    public getDocumentSettings(resource: string): Thenable<IJclSettings> {
        if (!this.capabilities.hasConfigurationCapability) {
            return Promise.resolve(this.globalSettings);
        }

        let result = this.documentSettings.get(resource);

        if (!result) {
            result = this.connection.workspace.getConfiguration({
                scopeUri: resource,
                section: "jclServer"
            });
            this.documentSettings.set(resource, result);
        }
        return result;
    }
}