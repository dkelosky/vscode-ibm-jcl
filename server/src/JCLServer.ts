import { InitializeParams, TextDocuments, InitializeResult, createConnection, ProposedFeatures, Connection, DidChangeConfigurationNotification, DocumentSymbolParams, SymbolInformation, TextDocumentChangeEvent, TextDocument, Diagnostic, DidChangeConfigurationParams, PublishDiagnosticsParams } from "vscode-languageserver";
import { Capabilities } from "./Capabilities";
import { Settings } from "./Settings";
import { Symbols } from "./Symbols";
import { Diagnostics } from "./Diagnostics";

export class JCLServer {

    public connection: Connection;
    public documents: TextDocuments;

    protected capabilities: Capabilities;
    protected settings: Settings;

    constructor() {

        // capabilities
        this.capabilities = new Capabilities();

        // Create a simple text document manager. The text document manager
        // supports full document sync only
        this.documents = new TextDocuments();

        // Create a connection for the server. The connection uses Node's IPC as a transport.
        // Also include all preview / proposed LSP features.
        this.connection = createConnection(ProposedFeatures.all);

        // settings
        this.settings = new Settings(this.connection, this.capabilities);

        // set listeners
        this.setDocumentsListeners();
        this.setConnectionListeners();

        // Make the text document manager listen on the connection
        // for open, change and close text document events
        this.documents.listen(this.connection);

        // listen on the connection
        this.connection.listen();
    }

    protected initialize(params: InitializeParams): InitializeResult {

        // set capabilities
        this.capabilities.init(params);

        return {
            capabilities: {
                textDocumentSync: this.documents.syncKind,
                documentSymbolProvider: true,
            }
        };
    }

    protected initialized(): void {
        if (this.capabilities.hasConfigurationCapability) {
            // Register for all configuration changes.
            this.connection.client.register(DidChangeConfigurationNotification.type, undefined);
        }
        if (this.capabilities.hasWorkspaceFolderCapability) {
            this.connection.workspace.onDidChangeWorkspaceFolders(() => {
                this.connection.console.error("Workspace folder change event received.");
            });
        }
    }

    protected didChangeConfiguration(change: DidChangeConfigurationParams): void {
        if (this.capabilities.hasConfigurationCapability) {
            // Reset all cached document settings
            this.settings.documentSettings.clear();
        } else {
            this.settings.globalSettings =
                (change.settings.languageServerExample || Settings.defaultSettings);
        }

        // Revalidate all open text documents
        this.documents.all().forEach(this.validateTextDocument.bind(this));
    }

    protected documentSymbol(parm: DocumentSymbolParams): SymbolInformation[] | null {
        const document = this.documents.get(parm.textDocument.uri);

        if (!document) {
            return null;
        }

        return Symbols.parse(document);
    }

    protected documentsDidClose(change: TextDocumentChangeEvent): void {
        // Only keep settings for open documents
        this.settings.documentSettings.delete(change.document.uri);
    }

    // The content of a text document has changed. This event is emitted
    // when the text document first opened or when its content has changed.
    protected documentsDidChangeContent(change: TextDocumentChangeEvent): void {
        this.validateTextDocument(change.document);
    }

    protected setDocumentsListeners(): void {
        this.documents.onDidClose(this.documentsDidClose.bind(this));
        this.documents.onDidChangeContent(this.documentsDidChangeContent.bind(this));
    }

    protected setConnectionListeners(): void {
        this.connection.onInitialize(this.initialize.bind(this));
        this.connection.onInitialized(this.initialized.bind(this));
        this.connection.onDidChangeConfiguration(this.didChangeConfiguration.bind(this));
        this.connection.onDocumentSymbol(this.documentSymbol.bind(this));
    }

    protected async validateTextDocument(textDocument: TextDocument): Promise<void> {

        // In this simple example we get the settings for every validate run.
        // const settings = await this.settings.getDocumentSettings(textDocument.uri);

        const diagnostics: Diagnostic[] = [];

        Diagnostics.checkLengths(textDocument, diagnostics);

        // Send the computed diagnostics to VSCode.
        this.sendDiagnostics({ uri: textDocument.uri, diagnostics });

    }

    protected sendDiagnostics(params: PublishDiagnosticsParams): void {
        this.connection.sendDiagnostics(params);
    }
}
