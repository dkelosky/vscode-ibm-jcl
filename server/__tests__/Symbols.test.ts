import { Symbols } from "../src/Symbols"
import { TextDocument } from "vscode-languageserver";

describe("Symbols tests", () => {

    it("should have one symbol", () => {
        const file = "//HELLO\r\n";
        const document = TextDocument.create("/nowhere", "jcl", 1, file);
        expect(Symbols.parse(document)).toMatchSnapshot();
    });


    it("should have no symbols", () => {
        const file = " //HELLO\r\n";
        const document = TextDocument.create("/nowhere", "jcl", 1, file);
        expect(Symbols.parse(document)).toMatchSnapshot();
    });
});
