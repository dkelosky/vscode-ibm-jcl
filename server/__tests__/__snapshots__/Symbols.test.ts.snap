// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Symbols tests should have no symbols 1`] = `Array []`;

exports[`Symbols tests should have one symbol 1`] = `
Array [
  Object {
    "kind": 14,
    "location": Object {
      "range": Object {
        "end": Object {
          "character": 6,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
      "uri": "/nowhere",
    },
    "name": "//HELLO",
  },
]
`;
